x/protorev/keeper/rebalance.go

```diff
@@ -167,30 +173,174 @@ func (k Keeper) FindMaxProfitForRoute(ctx sdk.Context, route RouteMetaData, rema
+		hop := route.Route[index]
+		pool, err := k.poolmanagerKeeper.GetPool(ctx, hop.PoolId)
+		if err != nil {
+			return sdk.ZeroInt(), err
+		}
+
+		tokenInDenom := inputDenom
+		if index > 0 {
+			tokenInDenom = route.Route[index-1].TokenOutDenom
+		}
+
+		switch {
+		case pool.GetType() == poolmanagertypes.Concentrated:
+			// If the pool is a concentrated liquidity pool, then check the maximum amount in that can be used
+			// and determine what this amount is as an input at the previous pool (working all the way up to the
+			// beginning of the route).
+			maxTokenIn, maxTokenOut, err := k.concentratedLiquidityKeeper.ComputeMaxInAmtGivenMaxTicksCrossed(
+				ctx,
+				pool.GetId(),
+				tokenInDenom,
+				types.MaxTicksCrossed,
+			)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+
+			// if there have been no other CL pools in the route, then we can set the intermediate coin to the max input amount.
+			// Additionally, if the amount of the previous token is greater than the possible amount from this pool, then we
+			// can set the intermediate coin to the max input amount (from the current pool). Otherwise we have to do a
+			// safe swap given the previous max amount.
+			if intermidiateCoin.IsNil() || maxTokenOut.Amount.LT(intermidiateCoin.Amount) {
+				intermidiateCoin = maxTokenIn
+				continue
+			}
+
+			// In the scenario where there are multiple CL pools in a route, we select the one that inputs
+			// the smaller amount to ensure we do not overstep the max ticks moved.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		case !intermidiateCoin.IsNil():
+			// If we have already seen a CL pool in the route, then simply propagate the intermediate coin up
+			// the route.
+			intermidiateCoin, err = k.executeSafeSwap(ctx, pool.GetId(), intermidiateCoin, tokenInDenom, route.StepSize)
+			if err != nil {
+				return sdk.ZeroInt(), err
+			}
+		}
+	}
+
+	// In the case where there are no CL pools, we want to return the extended max input amount
+	if intermidiateCoin.IsNil() {
+		return types.ExtendedMaxInputAmount, nil
+	}
+
+	// Ensure that the normalized upper bound is not greater than the extended max input amount
+	upperBound := intermidiateCoin.Amount.Quo(route.StepSize)
+	if upperBound.GT(types.ExtendedMaxInputAmount) {
+		return types.ExtendedMaxInputAmount, nil
+	}
+
+	return upperBound, nil
+}
+
+// executeSafeSwap executes a safe swap by first ensuring the swap amount is less than the
+// amount of total liquidity in the pool.
+func (k Keeper) executeSafeSwap(
+	ctx sdk.Context,
+	poolID uint64,
+	outputCoin sdk.Coin,
+	tokenInDenom string,
+	stepSize sdk.Int,
+) (sdk.Coin, error) {
+	liquidity, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolID)
+	if err != nil {
+		return sdk.NewCoin(tokenInDenom, sdk.ZeroInt()), err
+	}
+
+	// At most we can swap half of the liquidity in the pool
```

Comment:
This is not merge-blocking, but due to the way we handle rounding (always in the pool's favor and can be meaningfully high in some scenarios), this assumption does not necessarily true.

Rounding error can be nontrivial here and accumulates over time, so while this is likely the simplest way to handle this specific case, it should be kept in mind for future logic leaning on assumptions related to absolute assets in the pool.

Reply:
Yea, totally agree. We've been going back and forth on how to best estimate maximal swap in when CL/stable/balancer pools are involved and we more or less ran into the same issue with all of our approaches. I'll decrease this to `1/4` for now and will revisit this once we have a better idea of how frequently this is an actual issue. 

Also related, we plan on doing some structural refactoring - decoupling the various components of protorev - sometime soon which will let us navigate this with more granularity.

Reply:
@AlpinYukseloglu Based on your understanding, does the reduction to 1/4 improve the situation or are you thinking of a different approach here?